hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/client/utils/SparkMetadataWriterUtils.java

```diff
@@ -194,66 +207,21 @@ private static List<HoodieRecord> getBaseFileRecords(HoodieBaseFile baseFile, Ho
     }
   }
 
-  private static Dataset<Row> toDataset(List<HoodieRecord> records, Schema schema, SQLContext sqlContext, boolean isBaseFile) {
-    List<GenericRecord> avroRecords = records.stream()
+  private static List<Row> toRows(List<HoodieRecord> records, Schema schema, SQLContext sqlContext, String path) {
+    StructType structType = AvroConversionUtils.convertAvroSchemaToStructType(schema);
+    Function1<GenericRecord, Row> converterToRow = AvroConversionUtils.createConverterToRow(schema, structType);
+    List<Row> avroRecords = records.stream()
         .map(r -> {
-          if (isBaseFile) {
-            return (GenericRecord) r.getData();
-          }
-          HoodieRecordPayload payload = (HoodieRecordPayload) r.getData();
           try {
-            return (GenericRecord) payload.getInsertValue(schema).get();
+            return (GenericRecord) (r.getData() instanceof GenericRecord ? r.getData()
+                : ((HoodieRecordPayload) r.getData()).getInsertValue(schema, new Properties()).get());
           } catch (IOException e) {
-            throw new HoodieIOException("Failed to extract Avro payload", e);
+            throw new HoodieIOException("Could not fetch record payload");
           }
         })
+        .map(converterToRow::apply)
+        // .map(row -> RowFactory.create(path, row))
```

Comment:
can we remove L 223. (the commented out line)

Reply:
Addressed